src/interfaces/chain.h

```diff
@@ -56,74 +51,59 @@ class Chain
-
-        //! Get block median time past. Height must be valid or this function
-        //! will abort.
-        virtual int64_t getBlockMedianTimePast(int height) = 0;
-
-        //! Check that the block is available on disk (i.e. has not been
-        //! pruned), and contains transactions.
-        virtual bool haveBlockOnDisk(int height) = 0;
-
-        //! Return height of the first block in the chain with timestamp equal
-        //! or greater than the given time and height equal or greater than the
-        //! given height, or nullopt if there is no block with a high enough
-        //! timestamp and height. Also return the block hash as an optional output parameter
-        //! (to avoid the cost of a second lookup in case this information is needed.)
-        virtual Optional<int> findFirstBlockWithTimeAndHeight(int64_t time, int height, uint256* hash) = 0;
-
-        //! Return height of last block in the specified range which is pruned, or
-        //! nullopt if no block in the range is pruned. Range is inclusive.
-        virtual Optional<int> findPruned(int start_height = 0, Optional<int> stop_height = nullopt) = 0;
-
-        //! Return height of the specified block if it is on the chain, otherwise
-        //! return the height of the highest block on chain that's an ancestor
-        //! of the specified block, or nullopt if there is no common ancestor.
-        //! Also return the height of the specified block as an optional output
-        //! parameter (to avoid the cost of a second hash lookup in case this
-        //! information is desired).
-        virtual Optional<int> findFork(const uint256& hash, Optional<int>* height) = 0;
-
-        //! Get locator for the current chain tip.
-        virtual CBlockLocator getTipLocator() = 0;
-
-        //! Return height of the highest block on chain in common with the locator,
-        //! which will either be the original block used to create the locator,
-        //! or one of its ancestors.
-        virtual Optional<int> findLocatorFork(const CBlockLocator& locator) = 0;
-
-        //! Check if transaction will be final given chain height current time.
-        virtual bool checkFinalTx(const CTransaction& tx) = 0;
-    };
-
-    //! Return Lock interface. Chain is locked when this is called, and
-    //! unlocked when the returned interface is freed.
-    virtual std::unique_ptr<Lock> lock(bool try_lock = false) = 0;
+    //! Get current chain height, not including genesis block (returns 0 if
+    //! chain only contains genesis block, nullopt if chain does not contain
+    //! any blocks)
+    virtual Optional<int> getHeight() = 0;
+
+    //! Get block height above genesis block. Returns 0 for genesis block,
+    //! 1 for following block, and so on. Returns nullopt for a block not
+    //! included in the current chain.
+    virtual Optional<int> getBlockHeight(const uint256& hash) = 0;
+
+    //! Get block hash. Height must be valid or this function will abort.
```

Comment:
For getBlockHash, getBlockTime, and getBlockMedianTimePast, I don't think it safe anymore for these methods to abort when an invalid height is passed. Previously it was possible to guarantee that an invalid height would never be passed because `cs_main` would be held before calling these (and if an invalid height was passed it indicated a bug). But now that cs_main isn't held, there could be a race condition where `invalidateblock` is called, or there is a reorg to a shorter chain that would cause aborts now.

A possible fix would be to implement error handling for race conditions in the places where these methods are called, maybe changing these methods to throw an exception or return `Optional<>` values or success `bools` to indicate failure instead of aborting.

Another solution would be to drop these methods, and replace them with a more general method similar to [`findBlock`](https://github.com/bitcoin/bitcoin/blob/ddecb671f0a41eaa8e6198f7e55172c1256bd978/src/interfaces/chain.h#L128-L137) that is capable of returning multiple pieces of information atomically so there aren't new race conditions the wallet has to handle now because it isn't holding `cs_main`.

Reply:
You're right, even if you need bad-timed reorgs to trigger them. After thinking about this, I picked the option to return `Optional<>` because that's one giving us more control. Throwing exception don't fit everywhere where an absent block isn't fatal. Return boolean would mean to pass pointer to values which is a bit different than other interface methods (like `getHeight`). 

On dropping these methods, I think `getBlockTime` would be the easiest one if we can store block time in confirmation struct of transactions but we're going to be limited by not being able to serialize them. `getBlockMedianTimePast` and `getBlockHash` can go by refactoring rescan logic.

Reply:
I've been through function triplet callsites again and I think using hash instead of height wouldn't make code safer that much. 

`getBlockHash` is always called by rescan code which making the assumption of chain lock not being held.

`getBlockTime`/`getBlockMedianTimePast`, given than block time must always be higher than parent MTP and not to be more than 2h in the future by consensus rules and than wallet code is using time range to take decision it doesn't matter than height match with block hash A or block hash B.